### PR TITLE
Bluetooth: controller: fix post DLE/PHY update event length

### LINKED ISSUE #21995: Bluetooth: controller: split: Porting of connection event length
**Is your enhancement proposal related to a problem? Please describe.**
A clear and concise description of what the problem is.

Without use of connection event length, radio events would overlap and cause degradation in performance/throughput.

**Describe the solution you'd like**

Be inspired by zephyrproject-rtos/zephyr@32c2f85

**Describe alternatives you've considered**

**Additional context**
Add any other context or graphics (drag-and-drop an image) about the feature request here.

### LINKED ISSUE #52393: Controller: ACL packets NACKed after Data Length Update
**Describe the bug**
When central is connected to multiple peripherals and the ~MTU~ LL data length is increased to 247 bytes, central gets notifications only from the last connected peripheral

**To Reproduce**
1. Central connects to more than one peripheral
2. For each connection, set the connection interval to be 25 ms
3. For each connection, increase the MTU above 190 bytes
4. Send notifications from connected peripherals other than the last connected one

**Expected behavior**
1. Central receives notifications from the last connected peripheral
2. Notifications from peripherals other than the last connected one are missed
3. Peripherals that aren't communicating disconnect with time out disconnection reason

**Additional context**
Issue doesn't occur if the connection interval is increased or the notification data size is below 160 bytes

**Environment**
 - OS: Ubuntu 22.04
 - Toolchain : Zephyr SDK 0.15.1
 - Commit SHA 3607a05fbaf1d85a1e11d19c6ff33b869c5e88bb
 - PR https://github.com/zephyrproject-rtos/zephyr/pull/52392

### LINKED ISSUE #53049: Bluetooth: LL assertion fail with peripherals connect/disconnect rounds
**Describe the bug**
When central is connected to multiple peripherals and the MTU is increased to 247 bytes, central gets notifications only from the last connected peripheral

Central connects to 4 peripherals, exchange MTU to be increased to 247 bytes, then notifications should go in both directions for a minute before the peripherals disconnect and so on.

After several rounds, one of the LL assertions fails and operations stops

**To Reproduce**
1. Central connects to more than one peripheral
2. For each connection, set the connection interval to be 37 ms
3. Send notifications in both directions

Please checkout this commit f591f4116b2575017fc84bcdbd2182fc89b11584 and run the script ~/zephyrproject/zephyr/tests/bluetooth/bsim_bt/bsim_test_conn_stress/tests_scripts/conn_stress.sh

image

### PR DESCRIPTION
**Pull request by kruithofa**
Fix the controller implementation to perform connection event length reservation based on the completed Data Length Update and/or PHY Update Procedure.

Note: some optimisation is possible with regards to the length allocated, currently the maximum rx and tx time are used, which leaves a lot of unused airtime in case that the 2M phy is used. If necessary this will be fixed in a separate PR
fixes #21995
fixes #52393
Fixes #53049

Signed-off-by: Andries Kruithof <andries.kruithof@nordicsemi.no>

### CODE REVIEW

**Review by cvinayak — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/controller/ll_sw/lll_conn.h [hunk 1]
```diff
@@ -134,6 +136,11 @@ struct lll_conn {
 	struct ccm ccm_tx;
 #endif /* CONFIG_BT_CTLR_LE_ENC */
 
+#if defined(CONFIG_BT_CTLR_DATA_LENGTH) || defined(CONFIG_BT_CTLR_PHY)
+	uint8_t evt_len_upd:1;
+	uint8_t evt_len_adv:1;
```
[the remark this answers is not in the document]
Comment (kruithofa) [reply subsys/bluetooth/controller/ll_sw/lll_conn.h]: Variables renamed

File: subsys/bluetooth/controller/ll_sw/ull_conn.c [hunk 2]
```diff
@@ -1841,9 +1846,57 @@ void ull_conn_done(struct node_rx_event_done *done)
...
+
+		lll->evt_len_upd = 0;
+#if defined(CONFIG_BT_CTLR_PHY)
+		ready_delay = (lll->role) ?
+			lll_radio_rx_ready_delay_get(lll->phy_rx, 1) :
+			lll_radio_tx_ready_delay_get(lll->phy_tx, lll->phy_flags);
+#if defined(CONFIG_BT_CTLR_DATA_LENGTH)
+		tx_time = lll->dle.eff.max_tx_time;
+		if (lll->evt_len_adv) {
+			rx_time = lll->dle.remote.max_rx_time;
```
[the remark this answers is not in the document]
Comment (kruithofa) [reply subsys/bluetooth/controller/ll_sw/ull_conn.c]: Updated

File: subsys/bluetooth/controller/ll_sw/ull_conn.c [hunk 3]
```diff
@@ -1841,9 +1846,57 @@ void ull_conn_done(struct node_rx_event_done *done)
...
+/* Start of PR */
+#if !defined(CONFIG_BT_LL_SW_LLCP_LEGACY)
+#if defined(CONFIG_BT_CTLR_DATA_LENGTH) || defined(CONFIG_BT_CTLR_PHY)
+	if (lll->evt_len_upd) {
+		uint32_t ready_delay, rx_time, tx_time, ticks_slot;
+
+		lll->evt_len_upd = 0;
+#if defined(CONFIG_BT_CTLR_PHY)
+		ready_delay = (lll->role) ?
+			lll_radio_rx_ready_delay_get(lll->phy_rx, 1) :
```
[the remark this answers is not in the document]
Comment (kruithofa) [reply subsys/bluetooth/controller/ll_sw/ull_conn.c]: done

File: subsys/bluetooth/controller/ll_sw/ull_conn.c [hunk 4]
```diff
@@ -1841,9 +1846,57 @@ void ull_conn_done(struct node_rx_event_done *done)
...
+		if (lll->evt_len_adv) {
+			rx_time = lll->dle.remote.max_rx_time;
+		} else {
+			rx_time = lll->dle.eff.max_rx_time;
+		}
+#else /* CONFIG_BT_CTLR_DATA_LENGTH */
+		tx_time = MAX(PDU_MAX_US(PDU_DC_PAYLOAD_SIZE_MIN, 0, 0),
+			      PDU_MAX_US(PDU_DC_PAYLOAD_SIZE_MIN, 0, lll->phy_tx));
+		rx_time = MAX(PDU_MAX_US(PDU_DC_PAYLOAD_SIZE_MIN, 0, 0),
+			      PDU_MAX_US(PDU_DC_PAYLOAD_SIZE_MIN, 0, lll->phy_rx));
```
[the remark this answers is not in the document]
Comment (kruithofa) [reply subsys/bluetooth/controller/ll_sw/ull_conn.c]: done

File: subsys/bluetooth/controller/ll_sw/ull_conn.c [hunk 5]
```diff
@@ -1841,9 +1846,57 @@ void ull_conn_done(struct node_rx_event_done *done)
...
+			lll_radio_rx_ready_delay_get(0, 0) :
+			lll_radio_tx_ready_delay_get(0, 0);
+		tx_time = PDU_MAX_US(lll->dle.eff.max_tx_octets, 0, 0);
+		if (lll->evt_len_adv) {
+			rx_time = PDU_MAX_US(lll->dle.remote.max_rx_octets, 0, 0);
+		} else {
+			rx_time = PDU_MAX_US(lll->dle.eff.max_rx_octets, 0, 0);
+		}
+#endif /* CONFIG_BT_CTLR_PHY */
+		ticks_slot = HAL_TICKER_US_TO_TICKS(OFFS_UNIT_300_US + ready_delay + EVENT_IFS_US + rx_time + tx_time + 4);
```
[the remark this answers is not in the document]
Comment (kruithofa) [reply subsys/bluetooth/controller/ll_sw/ull_conn.c]: done

**Comment by erbr-ot** [comment]
A thought - couldn't we know (ie remember) the sum of current ready_delay and tx/rx time contributions, so that we only need to calc the difference (to get the slot_plus/minus) to perform the ticker update? Instead of doing the 'full monty'?

**Comment by kruithofa** [comment]
> A thought - couldn't we know (ie remember) the sum of current ready_delay and tx/rx time contributions, so that we only need to calc the difference (to get the slot_plus/minus) to perform the ticker update? Instead of doing the 'full monty'?

We could, but it is a trade-off between memory usage and processor time. These calculations are only done occasionally, so no big effect on processing time

**Comment by carlescufi** [comment]
@cvinayak please take another look

File: subsys/bluetooth/controller/ll_sw/ull_conn.c [hunk 6]
```diff
@@ -1874,8 +1929,8 @@ void ull_conn_done(struct node_rx_event_done *done)
 		ticker_status = ticker_update(TICKER_INSTANCE_ID_CTLR,
 					      TICKER_USER_ID_ULL_HIGH,
 					      ticker_id,
-					      ticks_drift_plus,
-					      ticks_drift_minus, 0, 0,
+					      ticks_drift_plus, ticks_drift_minus,
+					      ticks_slot_plus, ticks_slot_minus,
```
Comment (mtpr-ot) [inline subsys/bluetooth/controller/ll_sw/ull_conn.c:1933]: In my perspective, updating the slot reservation based on DLE `max_octets` is incorrect for this controller. The ticker shall only reserve the time needed for basic LL transactions, i.e. for a PDU size of 27 bytes in each direction.
Using `max_octets` will cause too heavy reservations for ACL connections, causing substantial scheduling collisions.

**Review by mtpr-ot — DISMISSED** [review]

File: subsys/bluetooth/controller/ll_sw/ull_conn.c [hunk 7]
```diff
@@ -1865,8 +1871,57 @@ void ull_conn_done(struct node_rx_event_done *done)
...
+						    rx_time +
+						    tx_time +
+						    4);
+		if (ticks_slot > conn->ull.ticks_slot) {
+			ticks_slot_plus = ticks_slot - conn->ull.ticks_slot;
+		} else {
+			ticks_slot_minus = conn->ull.ticks_slot - ticks_slot;
+		}
+		conn->ull.ticks_slot = ticks_slot;
+		printk("Ticks_slot after: %d\n", ticks_slot);
```
Comment (ahmedmoheb-nordic) [inline subsys/bluetooth/controller/ll_sw/ull_conn.c:1913]: Is that print message intended to be merged?

File: subsys/bluetooth/controller/Kconfig.ll_sw_split [hunk 8]
```diff
@@ -459,6 +459,14 @@ config BT_CTLR_CENTRAL_SPACING
 	  (active clock jitter) + 17040 (PDU rx) = (radio event overheads +
 	  34234) microseconds.
 
+config BT_CTLR_CONN_EVENT_LEN_UPDATE
+	bool "Update event length reservation after PHY or DLE update"
+	depends on !BT_LL_SW_LLCP_LEGACY && (BT_CTLR_DATA_LENGTH || BT_CTLR_PHY)
+	default n
```
Comment (mtpr-ot) [inline subsys/bluetooth/controller/Kconfig.ll_sw_split:465]: I think the recommendation is to not set `default n`, as it is in fact default. Also I'm wondering if the name should reflect that we're talking about slot reservation?

Comment (kruithofa) [reply subsys/bluetooth/controller/Kconfig.ll_sw_split:465]: You're probably right, a few issues popped up recently that will be fixed by this PR
As for the name, I am open to suggestions, or how about BT_CTLR_SLOT_RESERVATION_UPDATE

File: subsys/bluetooth/controller/ll_sw/ull_conn.c (hunk 7 above)
Comment (kruithofa) [reply subsys/bluetooth/controller/ll_sw/ull_conn.c:1913]: good catch

**Comment by carlescufi** [comment]
@mtpr-ot @cvinayak please review

**Review by cvinayak — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Review by mtpr-ot — COMMENTED** [review]
LGTM

**Pull request merged**